### ISSUE TITLE
Killed mutants move to covered when adding unrelated tests

### ISSUE THREAD

**Issue by Edgpaez** [stryker-mutator/stryker-js#3012]
First of all **thank you** to all maintainers and contributors of this package. I hope this is actually a bug and I'm not wasting anyone's time.
Also, this seems to be related to #2989

**Summary**
Killed mutants move to covered when I add more tests for another class.

Initially I have a simple class that assigns some instance fields based on inputs on the constructor:
```javascript
class Frame {
  constructor(type) {
    this.state = Frame.STATES.IN_PLAY;
    this.type = type || Frame.TYPES.REGULAR;
    this.score = null;
    this.firstThrow = null;
    this.secondThrow = null;
  }
}
```
Then I have a simple test for it:
```javascript
  it("starts as a regular frame by default. could be started as a Fill Ball", () => {
    const regularFrame = new Frame();
    expect(regularFrame.type).toEqual(Frame.TYPES.REGULAR);

    const fillFrame = new Frame(Frame.TYPES.FILL_BALL_FOR_SPARE);
    expect(fillFrame.type).toEqual(Frame.TYPES.FILL_BALL_FOR_SPARE);
  });
```
Running stryker shows me I cover all mutants related to the constructor: 
```
🎯 Killed: 9 FILL_BALL_FOR_SPARE: "", (src/frame.js:13:26) (and 4 more) ☂️ Covered 4 mutants
🎯 9 FILL_BALL_FOR_SPARE: "", (src/frame.js:13:26)
🎯 12 this.type = false; (src/frame.js:18:17)
🎯 11 this.type = true; (src/frame.js:18:17)
🎯 10 constructor(type) {} (src/frame.js:16:21)
🎯 13 this.type = type && Frame.TYPES.REGULAR; (src/frame.js:18:17)
```
So far so good.

But then, as I implemented more parts of the problem, I added another class and another test for that class. 
Now, running stryker shows me this:
```
☂️ Covered 4 mutants (yet didn't kill any of them)
☂️ 61 constructor(type) {} (src/frame.js:16:21)
☂️ 62 this.type = true; (src/frame.js:18:17)
☂️ 63 this.type = false; (src/frame.js:18:17)
☂️ 64 this.type = type && Frame.TYPES.REGULAR; (src/frame.js:18:17)
```
Naturally I didn't change anything in the original code from above. Just a new `bowling.spec.js` and a `bowling.js` file

The weirdest part is that if I remove some skipped tests in the new `bowling.spec.js` file, I get a different report on mutants from my original 'frame.spec.js':
```
🎯 Killed: 62 this.type = true; (src/frame.js:18:17) ☂️ Covered 4 mutants
🎯 62 this.type = true; (src/frame.js:18:17)
☂️ 61 constructor(type) {} (src/frame.js:16:21)
☂️ 63 this.type = false; (src/frame.js:18:17)
☂️ 64 this.type = type && Frame.TYPES.REGULAR; (src/frame.js:18:17)
```
The result varies from execution to execution, but always some mutants are covered and other killed.

I have attached a very simple repository (with only the four files implicated) with the following commits:
- b7f0bf32131094ea63ecaa59afd07c79a6ec583f  the code where all mutants are killed
- a51fef0b2d01a6d7b28cf5b841c5f1319ff9993c the code where all mutants are covered but not killed
- fdc0c55ea5f5a27edebb115dfd2f3f861686323f the code where some mutants are killed but others survive

The relevant test is at line 4 of `src/frame.spec.js`
[reduced-bowling.zip](https://github.com/stryker-mutator/stryker-js/files/6852117/reduced-bowling.zip)

**Stryker config**
```json
{
  "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
  "packageManager": "yarn",
  "reporters": [
    "html",
    "clear-text",
    "progress"
  ],
  "testRunner": "jest",
  "coverageAnalysis": "perTest"
}
```

**Test runner config**
```json
{
"jest": {
    "modulePathIgnorePatterns": [
      "package.json"
    ]
  }
}
```

**Stryker environment**
```
├─ @stryker-mutator/api@5.2.2
├─ @stryker-mutator/core@5.2.2
│  ├─ @stryker-mutator/api@5.2.2
│  ├─ @stryker-mutator/instrumenter@5.2.2
│  ├─ @stryker-mutator/util@5.2.2
├─ @stryker-mutator/instrumenter@5.2.2
│  ├─ @stryker-mutator/api@5.2.2
│  ├─ @stryker-mutator/util@5.2.2
├─ @stryker-mutator/jest-runner@5.2.2
│  ├─ @stryker-mutator/api@5.2.2
│  ├─ @stryker-mutator/util@5.2.2
├─ @stryker-mutator/util@5.2.2
```

```
jest@26.6.3
```

**Test runner environment**

```shell
yarn test
```

**Your Environment**

| software         | version(s)
| ---------------- | -------
| node             | v14.17.1 |
| yarn              |  1.22.10 |
| Operating System | Big Sur 11.4 |

**Add stryker.log**

I added two stryker.log:  
- [some-killed-some-covered-stryker.log](https://github.com/stryker-mutator/stryker-js/files/6852152/some-killed-some-covered-stryker.log) where some mutants are covered and some are killed 
- [all-covered-stryker.log](https://github.com/stryker-mutator/stryker-js/files/6852146/all-covered-stryker.log) where when all are covered (none are killed)

I'd be happy to help with further debugging.

**Comment by nicojs** [comment]
I doubt it's related to #2989, since I figured out that one is related to the karma runner 😌

I'll try to take a look at this one next week

**Comment by Edgpaez** [comment]
I now believe this is related to Stryker bailing on the first test that kills a mutant. Let me elaborate:

Digging deeper I found in the `stryker.log` that the mutants are actually killed. 

For example:
```
15:24:14 (62008) DEBUG ClearTextReporter #61. [Killed] BlockStatement
15:24:14 (62008) DEBUG ClearTextReporter [36msrc/frame.js[39m:[33m16[39m:[33m21[39m
15:24:14 (62008) DEBUG ClearTextReporter [31m-     constructor(type) {[39m
15:24:14 (62008) DEBUG ClearTextReporter [31m-       this.state = Frame.STATES.IN_PLAY;[39m
15:24:14 (62008) DEBUG ClearTextReporter [31m-       this.type = type || Frame.TYPES.REGULAR;[39m
15:24:14 (62008) DEBUG ClearTextReporter [31m-       this.score = null;[39m
15:24:14 (62008) DEBUG ClearTextReporter [31m-       this.firstThrow = null;[39m
15:24:14 (62008) DEBUG ClearTextReporter [31m-       this.secondThrow = null;[39m
15:24:14 (62008) DEBUG ClearTextReporter [31m-     }[39m
15:24:14 (62008) DEBUG ClearTextReporter [32m+     constructor(type) {}[39m
15:24:14 (62008) DEBUG ClearTextReporter Killed by: Bowling receives plays
```

In the Mutants tab of the HTML report I see that all mutants  from the constructor were killed:
![image](https://user-images.githubusercontent.com/535862/126916760-144d6510-ad0f-4b88-9e45-6149d01bb85e.png)

However, on the Tests tab of the HTML report I see that the test 32 did not kill the mutants, only covered them:
![image](https://user-images.githubusercontent.com/535862/126916789-44ec0291-1078-4ec1-802b-99f005b296ae.png)

I also extracted the `MutationTestResult` from the HTML and confirmed that the mutants are have `status: "Killed"`. However, I also noticed this: `killedBy: ["0"],` 
[pretty-covered-not-killed-mutation-test-result.txt](https://github.com/stryker-mutator/stryker-js/files/6874750/covered-not-killed-MutationTestResult.txt)


**What I think is happening is:**
- test run in a particular order so that a test different than 32 fails and kills the mutants in question.
- Stryker bails after the first test killing the mutant
- when building the Tests tab, the report knows that the Mutants in the constructor (61, 62, 63 and 64) are covered by my test 32. But the one that killed it was 0 (or whatever previous test that had failed), so it reports that test 32 only covered the mutants.


That, I would argue, would be wrong since test 32 actually kills the mutants, stryker just bailed before finding that out.

Maybe the solution is along the lines of what @nicojs proposed [here](https://github.com/stryker-mutator/stryker-js/issues/2996#issuecomment-884298880)

Would that be correct?

**Comment by nicojs** [comment]
Ah yes, of course. Sorry, didn't realize right away that this was about the test view. You are correct.

It seems that you have a generic test killing a bunch of mutants at once which is executed first. It basically does to many assertions. You than have more tests for each scenario in isolation, but those are not responsible for killing additional mutants, hence the `Covering` test state.

I've thought about this a lot during the implementation and personally like this behavior. It shouldn't happen if one practices strict TDD. Having too many assertions is also considered to be a bad practice, because it makes for brittle tests that are hard to refactor.

That being said, I can see some users would prefer to see the full list of 'killed mutants', which is part of the reason I'm willing to make this behavior configurable.

**Comment by Edgpaez** [comment]
I think it is not a matter of strictly following TDD but the either doing TDD as "mockist" or in the "classical" sense as [defined by Martin Fowler](https://martinfowler.com/articles/mocksArentStubs.html#ClassicalAndMockistTesting).

So, in my example I did TDD. I just did not mock everything for each test. So, naturally when I wrote other tests they would cover some of the dependencies of the sut, and the respective mutants, letting the test runner decide which test might take the credit for killing the mutants.

I would argue that the docs (which are very good now) could do with a brief explanation of how to read/use the html report so as to reduce this possible confusion. Given the popularity of "classical" TDD it might be that I'm not the only one affected by this.

I will give both the documentation and the new behavior a try if you don't mind and would be happy to receive some guidance/suggestions.

**Comment by nicojs** [comment]
Yeah, the test view is a [fairly new addition](https://stryker-mutator.io/blog/announcing-stryker-js-5-brave-new-dimension.md/). T.b.h, I didn't even know how useful it would be :). 

> I will give both the documentation and the new behavior a try if you don't mind and would be happy to receive some guidance/suggestions.

That would be awesome! Just to be clear, you mean a PR to improve the docs and a PR to implement https://github.com/stryker-mutator/stryker-js/issues/2996#issuecomment-884298880 ? 😍

Feel free to ask any questions you might have. The way we're developing is documented in CONTRIBUTING.md. You can also reach me on slack: https://join.slack.com/t/stryker-mutator/shared_invite/enQtOTUyMTYyNTg1NDQ0LTU4ODNmZDlmN2I3MmEyMTVhYjZlYmJkOThlNTY3NTM1M2QxYmM5YTM3ODQxYmJjY2YyYzllM2RkMmM1NjNjZjM

**Comment by nicojs** [comment]
Support for `--disableBail` was merged in #3074 and released in StrykerJS v5.4
